### PR TITLE
handle multiple json objects in a message

### PR DESCRIPTION
**Pull request by BrunoQuaresma**
JIRA Ticket: https://faunadb.atlassian.net/browse/ENG-2374

### CODE REVIEW

File: src/_json.js [hunk 1]
```diff
@@ -16,6 +16,36 @@ function parseJSON(json) {
...
+
+  try {
+    results.push(parseJSON(json))
+  } catch (error) {
+    var position
+
+    /** JSON parse error message is throw by the browser */
+    if (error.message.match(/JSON.parse/g)) {
+      var matchResult = error.message.match(/column ([0-9+]*)/)
+      position = matchResult[1] - 1
```
Comment (faunaee) [inline src/_json.js:30]: We should avoid multiple regex operations where possible; this code could get called very frequently.

```suggestion
    var capture
    if (capture = error.message.match(/JSON.parse: unexpected non-whitespace.*column (\d+)/)) {
      position = capture[1] - 1
```

File: src/_json.js [hunk 2]
```diff
@@ -16,6 +16,36 @@ function parseJSON(json) {
...
+      /** Unexpected token message is throw by NodeJS */
+    } else if (error.message.match(/Unexpected token/g)) {
+      var matchResult = error.message.match(/at position ([0-9+]*)$/)
+      position = matchResult[1]
```
Comment (faunaee) [inline src/_json.js:34]: Use the same strategy here.

Note that your capture group syntax seems off here. `\d` matches digits. I'm not sure why your character array includes `+`, nor why would accept 0 or more digits.

The capture group in my suggestion above seems better.

File: src/_json.js [hunk 3]
```diff
@@ -16,6 +16,36 @@ function parseJSON(json) {
...
+    } else {
+      throw error
+    }
+
+    var left = json.slice(0, position)
+    var right = json.slice(position)
```
Comment (faunaee) [inline src/_json.js:40]: `slice` is for arrays. Why don't we use `substr` here? Then you don't have to do position math (above).

**Review by faunaee — CHANGES_REQUESTED** [review]
A few comments on the implementation.

The new `parseJSONStreaming` function should definitely have unit tests covering single and multiple JSON objects.

File: src/_json.js (hunk 1 above)
Comment (danieltodonnell) [reply src/_json.js:30]: In a browser at least, `.indexOf()` is 10 times faster than a regex, so I would suggest this instead.  We test FOR the error quickly, then use the more expensive operation to get the position if it's happened

if (error.message.indexOf('JSON.parse')) {
      var matchResult = error.message.match(/column ([0-9+]*)/)

File: src/_json.js (hunk 2 above)
Comment (danieltodonnell) [reply src/_json.js:34]: Likewise, i think indexOf will be more efficient.

File: src/_json.js (hunk 1 above)
Comment (danieltodonnell) [reply src/_json.js:30]: https://www.measurethat.net/Benchmarks/Show/975/11/regex-vs-indexof-vs-startswith-vs-substr#latest_results_block

File: src/_json.js (hunk 2 above)
Comment (danieltodonnell) [reply src/_json.js:34]: Ewan is correct about the regex.. change it to `/at position ([0-9]+)/`

**Review by danieltodonnell — CHANGES_REQUESTED** [review]

File: src/_json.js (hunk 1 above)
Comment (danieltodonnell) [reply src/_json.js:30]: This regex is incorrect as well

use `/column ([0-9]+)/`

Comment (BrunoQuaresma) [reply src/_json.js:30]: > This regex is incorrect as well
> 
> use `/column ([0-9]+)/`

It is not. `column ([0-9]+)` didn't work. 😕

Comment (BrunoQuaresma) [reply src/_json.js:30]: Ok, I'm going to use `indexOf` to check the error.

**Comment by BrunoQuaresma** [comment]
> The new `parseJSONStreaming` function should definitely have unit tests covering single and multiple JSON objects.

I totally agree with this 😄 . I just wanted to collect some feedback first.

File: src/_json.js (hunk 2 above)
Comment (BrunoQuaresma) [reply src/_json.js:34]: The same of before, `/at position ([0-9]+)/` didn't work.

File: src/_json.js (hunk 3 above)
Comment (BrunoQuaresma) [reply src/_json.js:40]: I think for this case there is no diff, and String also has a `slice()` method.

https://stackoverflow.com/questions/2243824/what-is-the-difference-between-string-slice-and-string-substring

File: src/_json.js (hunk 1 above)
Comment (faunaee) [reply src/_json.js:30]: @danieltodonnell How is the regex that I suggested incorrect? I tested it against the error message quoted in the issue description. I might be overlooking regex support in browsers that the driver supports, so I'm curious what I'm missing.

Comment (erickpintor) [reply src/_json.js:30]: Are we sure the message is stable across browsers + node js versions?

**Review by erickpintor — COMMENTED** [review]
For safely, we should also set the text decoder to streaming mode.

File: src/_json.js (hunk 1 above)
Comment (BrunoQuaresma) [reply src/_json.js:30]: @erickpintor I was only able to catch these 2 messages but I'm not 100% sure these are the only possible messages.

**Comment by BrunoQuaresma** [comment]
Interesting discovery: Firefox JS error messages are diff from Chrome that are diff from Safari and go on.

Screen Shot 2020-10-26 at 09 16 16
Screen Shot 2020-10-26 at 09 14 38
Screen Shot 2020-10-26 at 09 14 24

I think we can move forward with the current solution since this issue only happens, from what we know, in Firefox.

**Comment by erickpintor** [comment]
https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API

image

It does seem like a firefox only issue. Would you mind adding some notes in the code for it?

File: src/stream.js [hunk 4]
```diff
@@ -156,15 +156,19 @@ StreamClient.prototype.subscribe = function() {
...
-      self._client.syncLastTxnTime(event.txnTS)
-    }
-    if (event.event === 'error') {
-      onError(new errors.StreamErrorEvent(event))
-    } else {
-      self._onEvent(event)
-    }
+    var events = json.parseJSONStreaming(data)
+
+    events.forEach(function(event) {
```
Comment (erickpintor) [inline src/stream.js:161]: Is `forEach` ES5 compatible?

File: test/_json.test.js [hunk 5]
```diff
@@ -0,0 +1,11 @@
+const { parseJSONStreaming } = require('../src/_json')
+
+describe('parseJSONStreaming', () => {
+  test('mutiple JSON objects in a message', () => {
+    // const jsonText = `{\"event\":\"version\",\"txnTS\":1603455982860000,\"data\":{\"ref\":{\"@ref\":{\"id\":\"278763355019149825\",\"collection\":{\"@ref\":{\"id\":\"Status\",\"collection\":{\"@ref\":{\"id\":\"collections\"}}}}}},\"ts\":{\"@ts\":\"2020-10-23T12:26:22.860Z\"},\"action\":\"update\",\"new\":{\"data\":{\"game\":1,\"players\":3,\"scores\":[305,27,9]}}}}{\"event\":\"version\",\"txnTS\":1603455983170000,\"data\":{\"ref\":{\"@ref\":{\"id\":\"278763355019149825\",\"collection\":{\"@ref\":{\"id\":\"Status\",\"collection\":{\"@ref\":{\"id\":\"collections\"}}}}}},\"ts\":{\"@ts\":\"2020-10-23T12:26:23.170Z\"},\"action\":\"update\",\"new\":{\"data\":{\"game\":1,\"players\":3,\"scores\":[306,27,9]}}}}`
```
Comment (erickpintor) [inline test/_json.test.js:5]: What is this comment for?

File: src/stream.js (hunk 4 above)
Comment (BrunoQuaresma) [reply src/stream.js:161]: yes, it is.

File: test/_json.test.js (hunk 5 above)
Comment (BrunoQuaresma) [reply test/_json.test.js:5]: Nice catch. Removing it.

**Comment by BrunoQuaresma** [comment]
> t does seem like a firefox only issue. Would you mind adding some notes in the code for it?

I added a note for it on top of the `parseJSONStreaming()` function.

```
/**
 * This function allow us to parse responses as streaming
 * currently, this is not allowed by Firefox only
 * https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API
 */
```

**Review by ryancharris — CHANGES_REQUESTED** [review]
CI is failing

**Comment by erickpintor** [comment]
Are those test failures legit or are we using the wrong build in the CI? Otherwise this change looks good to me.

**Comment by BrunoQuaresma** [comment]
@erickpintor on local, this error is failing but it is also failing on master: 'wraps error events'. Could u check if this is working on your local?

**Comment by BrunoQuaresma** [comment]
Since the error is also happening at master I'm merging this but I'm also investigating what is happening.

**Pull request merged**